applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.cpp

```diff
@@ -0,0 +1,294 @@
+    return rValue;
+}
+
+//************************************************************************************
+//************************************************************************************
+
+array_1d<double, 3 > & TrussPlasticityConstitutiveLaw::GetValue(
+    const Variable<array_1d<double, 3 > >& rThisVariable,
+    array_1d<double, 3 > & rValue)
+{
+    if (rThisVariable == FORCE)
+    {
+        rValue[0] = this->mStressState;
+        rValue[1] = 0.0;
+        rValue[2] = 0.0;
+    }
+    else KRATOS_ERROR << "can't get the specified value" << std::endl;
+    return rValue;
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void TrussPlasticityConstitutiveLaw::SetValue(
+    const Variable<double>& rThisVariable,
+    const double& rValue,
+    const ProcessInfo& rCurrentProcessInfo
+    )
+{
+    if(rThisVariable == PLASTIC_STRAIN) this->mAccumulatedPlasticStrain = rValue;
+    if(rThisVariable == PLASTIC_ALPHA) this->mPlasticAlpha = rValue;
+    else KRATOS_ERROR << "can't set the specified value" << std::endl;
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void TrussPlasticityConstitutiveLaw::SetValue(const Variable<bool>& rThisVariable,
+                          const bool& rValue,
+                          const ProcessInfo& rCurrentProcessInfo)
+{
+    if(rThisVariable == INELASTIC_FLAG) this->mInelasticFlag = rValue;
+    else KRATOS_ERROR << "can't set the specified value" << std::endl;
+}
+
+
+//************************************************************************************
+//************************************************************************************
+
+double& TrussPlasticityConstitutiveLaw::CalculateValue(
+    ConstitutiveLaw::Parameters& rParameterValues,
+    const Variable<double>& rThisVariable,
+    double& rValue
+    )
+{
+    if(rThisVariable == TANGENT_MODULUS)
+    {    
+        const Properties& r_material_properties = rParameterValues.GetMaterialProperties(); 
+        const double hardening_modulus = r_material_properties[HARDENING_MODULUS_1D];
+        const double youngs_modulus = r_material_properties[YOUNG_MODULUS];
+        if (this->mInelasticFlag) 
+        {
+            rValue = (hardening_modulus*youngs_modulus)/(hardening_modulus+youngs_modulus);;
+        }
+        else rValue = youngs_modulus;
+    } 
+    else KRATOS_ERROR << "can't calculate the specified value" << std::endl;
+
+    return rValue;
+}
+
+Vector& TrussPlasticityConstitutiveLaw::CalculateValue(
+    ConstitutiveLaw::Parameters& rParameterValues,
+    const Variable<Vector>& rThisVariable,
+    Vector& rValue)
+{
+    
+    if(rThisVariable == NORMAL_STRESS) 
+    {
+        const SizeType dofs = 6;
```

Comment:
Maybe this can be defined as a static constexpr, so it doesn't need to be redefined again

Reply:
it only appears at this position

Reply:
constexpr means that it is a copiletime-const